src/node_crypto.cc

```diff
@@ -3190,16 +3201,102 @@ void CipherBase::InitIv(const FunctionCallbackInfo<Value>& args) {
     iv_buf = Buffer::Data(args[2]);
     iv_len = Buffer::Length(args[2]);
   }
-  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len);
+  CHECK(args[3]->IsInt32());
+  // Don't assign to cipher->auth_tag_len_ directly; the value might not
+  // represent a valid length at this point.
+  int auth_tag_len = args[3]->Int32Value();
+
+  cipher->InitIv(*cipher_type, key_buf, key_len, iv_buf, iv_len, auth_tag_len);
+}
+
+
+bool CipherBase::InitAuthenticated(const char *cipher_type, int iv_len,
+                                   int auth_tag_len) {
+  CHECK(IsAuthenticatedMode());
+
+  // TODO(tniessen) Use EVP_CTRL_AEAD_SET_IVLEN when migrating to OpenSSL 1.1.0
+  static_assert(EVP_CTRL_CCM_SET_IVLEN == EVP_CTRL_GCM_SET_IVLEN,
+                "OpenSSL constants differ between GCM and CCM");
+  if (!EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_GCM_SET_IVLEN, iv_len, nullptr)) {
+    env()->ThrowError("Invalid IV length");
+    return false;
+  }
+
+  if (EVP_CIPHER_CTX_mode(ctx_) == EVP_CIPH_CCM_MODE) {
+    if (auth_tag_len < 0) {
+      char msg[128];
+      snprintf(msg, sizeof(msg), "authTagLength required for %s", cipher_type);
+      env()->ThrowError(msg);
+      return false;
+    }
+
+#ifdef NODE_FIPS_MODE
+    // TODO(tniessen) Support CCM decryption in FIPS mode
+    if (kind_ == kDecipher && FIPS_mode()) {
+      env()->ThrowError("CCM decryption not supported in FIPS mode");
+      return false;
+    }
+#endif
+
+    if (!EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_CCM_SET_TAG, auth_tag_len,
+        nullptr)) {
+      env()->ThrowError("Invalid authentication tag length");
+      return false;
+    }
+
+    // When decrypting in CCM mode, this field will be set in setAuthTag().
+    if (kind_ == kCipher)
+      auth_tag_len_ = auth_tag_len;
+
+    // The message length is restricted to 2 ^ (15 - iv_len) - 1 bytes.
+    CHECK(iv_len >= 7 && iv_len <= 13);
+    switch (iv_len) {
+    case 13:
+    case 12:
+#if INT_MAX >= 4294967295
+    case 11:
+#endif
+#if INT_MAX >= 1099511627775
+    case 10:
+#endif
+#if INT_MAX >= 281474976710655
+    case 9:
+#endif
+#if INT_MAX >= 72057594037927935
+    case 8:
+#endif
+#if INT_MAX >= 18446744073709551615
+    case 7:
+#endif
```

Comment:
I wonder if we need to check such a large size of plaintext. 
The string length in ES2016 is defined less than Number.MAX_SAFE_INTEGER (2^53-1) in http://www.ecma-international.org/ecma-262/7.0/#sec-ecmascript-language-types-string-type.
Actually in Node, the maximum length of string and  buffer are defined in https://github.com/nodejs/node/blob/c9794880e89dbbecbbd85e4ee0980ed9c7ac0971/deps/v8/include/v8.h#L2477
and https://github.com/nodejs/node/blob/c9794880e89dbbecbbd85e4ee0980ed9c7ac0971/src/node_buffer.h#L35
so that both are less than 32bits.
It seems to me that it is enough to check the plaintext length in case that `iv_len` is from 11 to 13. 
Did I miss something?

Reply:
I did not consider any practical limits, internally, the crypto API uses ints, so I thought it would be fair to cover all possible integer inputs. (`case 7` is debatable as it won't fit into a signed 64 bit integer.)

Reply:
Apart from that, does this change match your expectations?

Reply:
> Apart from that, does this change match your expectations?

Yes. I built this in my local machine and tests work fine.

Reply:
Can we solve this without listing all the statements?

Something like `if (a >= static_cast<int>(15.5 - log2(INT_MAX) / 8.0))`?  (15.5 because `log2(INT_MAX)` is never going to exactly be an integer, but very close to one, and the `static_cast is` going to round down anyway.) At least gcc can compile that down to a simple comparison against an integer constant.